### PR TITLE
Consolidate batch-routing errors into errors section

### PR DESCRIPTION
**Pull request by evansiroky**
This PR changes how error responses are shown in the narrative-itineraries (batch routing).

Screen Recording 2021-04-19 at 11 31 33 PM

The highlights are as follows:

- Consolidates all batch routing error messages into a list of errors
- Refactors some error display to share common code
- Has the ability to toggle between errors and plannable itineraries
- Has the ability to show feed-wide rental errors (requires https://github.com/ibi-group/OpenTripPlanner/pull/53).

### CODE REVIEW

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
I prefer this approach over the tooltip. Please consider #360, importing `Icon`, and completing JSDoc where mentioned.

File: lib/components/narrative/narrative-itineraries.js [hunk 1]
```diff
@@ -101,9 +118,168 @@ class NarrativeItineraries extends Component {
...
+      <>
+        {/* force a line break with css flex so the errors button show up
+          below the summary and sort selector */}
+        <div style={{ flexBasis: '100%', height: 5 }} />
+        <div>
+          {this._renderErrorsButton()}
+        </div>
+      </>
+    )
+  }
```
[the remark this answers is not in the document]
Comment (evansiroky) [reply lib/components/narrative/narrative-itineraries.js]: #360 has been merged into this branch

File: lib/components/narrative/narrative-itineraries.js [hunk 2]
```diff
@@ -101,9 +118,168 @@ class NarrativeItineraries extends Component {
...
+      content = (
+        <>
+          <button
+            className='clear-button-formatting'
+            onClick={itineraryIsExpanded
+              ? this._toggleDetailedItinerary
+              : this._toggleShowErrors
+            }
+          >
+            <i className='fa fa-arrow-left' /> View all options
```
[the remark this answers is not in the document]
Comment (evansiroky) [reply lib/components/narrative/narrative-itineraries.js]: Refactored in 85f1753

File: lib/components/narrative/narrative-itineraries.js [hunk 3]
```diff
@@ -101,9 +118,168 @@ class NarrativeItineraries extends Component {
...
+      )
+    })
+  }
+
+  _renderErrorsButton = () => {
+    const { errors } = this.props
+    if (errors.length === 0) return null
+    return (
+      <IssueButton onClick={this._toggleShowErrors}>
+        <i className='fa fa-warning' />
```
[the remark this answers is not in the document]
Comment (evansiroky) [reply lib/components/narrative/narrative-itineraries.js]: Refactored in 85f1753

File: lib/components/narrative/narrative-itineraries.js [hunk 4]
```diff
@@ -101,9 +118,168 @@ class NarrativeItineraries extends Component {
...
+    return errors.map((error, idx) => {
+      let icon
+      if (error.network) {
+        const CompanyIcon = getCompanyIcon(error.network)
+        icon = <CompanyIcon />
+      } else {
+        icon = <Icon className='text-warning' type='exclamation-triangle' />
+      }
+      return (
+        <div className='option default-itin' key={idx}>
```
[the remark this answers is not in the document]
Comment (evansiroky) [reply lib/components/narrative/narrative-itineraries.js]: Refactored in 85f1753

File: lib/util/state.js [hunk 5]
```diff
@@ -25,32 +25,142 @@ export function getTimestamp (time = moment()) {
...
+          networkRentalInfo
+        })
+      }
+    }
+  })
+
+  return networksWithFeedWideErrors
+}
+
+function humanReadableMode (modeStr) {
```
[the remark this answers is not in the document]
Comment (evansiroky) [reply lib/util/state.js:60]: Added in 85f1753

File: lib/util/state.js [hunk 6]
```diff
@@ -25,32 +25,142 @@ export function getTimestamp (time = moment()) {
...
+  if (!rentalInfo) return networksWithFeedWideErrors
+
+  // iterate through all networks to deter
+  Object.keys(rentalInfo).forEach(network => {
+    const networkRentalInfo = rentalInfo[network]
+    if (
+      networkRentalInfo.errors &&
+        networkRentalInfo.errors.length > 0
+    ) {
+      if (networkRentalInfo.errors.some(
```
[the remark this answers is not in the document]
Comment (evansiroky) [reply lib/util/state.js]: Refactored in 85f1753

File: lib/util/state.js [hunk 7]
```diff
@@ -25,32 +25,142 @@ export function getTimestamp (time = moment()) {
...
+
+  // get first available rental info (this assumes that there won't be
+  // 2 rental info types (for example car rental access and vehicle
+  // rental egress))
+  const rentalInfo = response.bikeRentalInfo ||
+    response.carRentalInfo ||
+    response.vehicleRentalInfo
+  if (!rentalInfo) return networksWithFeedWideErrors
+
+  // iterate through all networks to deter
```
[the remark this answers is not in the document]
Comment (evansiroky) [reply lib/util/state.js]: Refactored in 85f1753

**Review by binh-dam-ibigroup — APPROVED** [review]
Looks good. Aside, I made a comment on company chat regarding how we want to handle the repeat text when an itinerary is not found.

File: lib/util/state.js [hunk 8]
```diff
@@ -25,32 +25,136 @@ export function getTimestamp (time = moment()) {
...
+function humanReadableMode (modeStr) {
+  if (!modeStr) return 'N/A'
+  const arr = modeStr.toLowerCase().replace(/_/g, ' ').split(',')
+  if (arr.length > 2) {
+    const last = arr.pop()
+    return arr.join(', ') + ' and ' + last
+  } else {
+    return arr.join(' and ')
+  }
+}
```
Comment (landonreed) [inline lib/util/state.js:69]: Thoughts on replacing the second part of this method with https://www.npmjs.com/package/humanize-list?

File: lib/util/state.js [hunk 9]
```diff
@@ -25,32 +25,136 @@ export function getTimestamp (time = moment()) {
...
+
+/**
+ * Generates a list of issues with a list of OTP responses. For batch routing
+ * this could be an array of multiple responses, but for standard itinerary
+ * routing it will be at most an array with one response.
```
Comment (landonreed) [inline lib/util/state.js:74]: Is this jsdoc correct? Wouldn't the array length be greater than one for standard itin routing if there was an OTP response error AND a feed-wide rental error?

File: lib/util/state.js [hunk 10]
```diff
@@ -25,32 +25,136 @@ export function getTimestamp (time = moment()) {
...
- * Gets the active error (first OTP plan response that contains an error). Only
- * one OTP response is expected in itinerary routing, but if batch routing is
- * enabled, multiple responses with errors should be handled.
+ * Gets the active error (first OTP plan response that contains an error). This
+ * should only be used with itinerary routing, not batch routing.
  */
 export function getActiveError (otpState) {
-  const errorResponse = getResponsesWithErrors(otpState)[0]
-  if (errorResponse && errorResponse.error) return errorResponse.error
+  return getResponsesWithErrors(otpState)[0]
```
Comment (landonreed) [inline lib/util/state.js:129]: Will this now show just the feedWide rental error if the list of errors has a rental error + OTP response error? If so, is that the desired behavior?

File: lib/util/state.js [hunk 11]
```diff
@@ -25,32 +25,136 @@ export function getTimestamp (time = moment()) {
...
+}
+
+/**
+ * Get the appropriate message for a trip planning error given possibly pre-
+ * configured message overrides form the config.
```
Comment (landonreed) [inline lib/util/state.js:134]: Typo
```suggestion
 * configured message overrides from the config.
```

**Review by landonreed — APPROVED** [review]
Looks really good, @evansiroky! Just a few comments that need addressing before merging.

File: lib/util/state.js (hunk 8 above)
Comment (evansiroky) [reply lib/util/state.js:69]: Adding that package seems to add more lines of code overall. It's just 6 lines of code currently, so I don't think it's worth adding another package.

File: lib/components/narrative/narrative-itineraries.js [hunk 12]
```diff
@@ -101,9 +118,145 @@ class NarrativeItineraries extends Component {
     setUseRealtimeResponse({useRealtime: !useRealtime})
   }
 
+  _toggleShowErrors = () => {
+    this.setState({ showingErrors: !this.state.showingErrors })
+  }
+
+  _renderErrors = () => {
```
[the remark this answers is not in the document]
Comment (evansiroky) [reply lib/components/narrative/narrative-itineraries.js]: Well, I refactored it out in 8528d89154ec5e471d8ac40fc66ce4e35d974261, but the prop drilling is not pretty. If I had more time to make it look prettier, I'd move more stuff into actions and make the separate components connect to the redux store. Lmk if this is desired over all our other priorities.

File: lib/util/state.js (hunk 9 above)
Comment (evansiroky) [reply lib/util/state.js]: I updated the JSDoc accordingly in 8528d89154ec5e471d8ac40fc66ce4e35d974261

File: lib/util/state.js (hunk 10 above)
Comment (evansiroky) [reply lib/util/state.js]: Changed so it always gets the last error in 8528d89154ec5e471d8ac40fc66ce4e35d974261

**Review by landonreed — APPROVED** [review]
Looks good, but a couple of cosmetic things that need resolving:
- the save trip button isn't aligned correctly (see screenshot)
- the padding in the errors button is slightly off (right side needs a bit more padding to match the left and bottom seems slightly more padded than the top)

image

image

**Pull request merged**